File: day-01/part-2/youyoun.rb
def to_i_2(obj)
    if obj == "one" or obj == "1"
        return 1
    elsif obj == "two" or obj == "2"
        return 2
    elsif obj == "three" or obj == "3"
        return 3
    elsif obj == "four" or obj == "4"
        return 4
    elsif obj == "five" or obj == "5"
        return 5
    elsif obj == "six" or obj == "6"
        return 6
    elsif obj == "seven" or obj == "7"
        return 7
    elsif obj == "eight" or obj == "8"
        return 8
    elsif obj == "nine" or obj == "9"
        return 9
    end
end

def run(s)
    sum = 0
    for line in s
        matches = line.scan(/(?=([1-9]|one|two|three|four|five|six|seven|eight|nine))/)
        first = to_i_2(matches[0][0])
        last = to_i_2(matches[-1][0])

        sum += first * 10 + last
    end
    return sum
end



starting = Process.clock_gettime(Process::CLOCK_MONOTONIC)
answer = run(ARGV[0].lines)
elapsed = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - starting) * 1000

puts "_duration:#{elapsed}\n#{answer}"

File: day-01/part-2/youyoun.rs
use std::collections::HashMap;
use std::env::args;
use std::time::Instant;

fn main() {
    let now = Instant::now();
    let output = run(&args().nth(1).expect("Please provide an input"));
    let elapsed = now.elapsed();
    println!("_duration:{}", elapsed.as_secs_f64() * 1000.);
    println!("{}", output);
}

struct Node {
    children: HashMap<char, Node>,
    is_end: bool,
    number: u32,
}

impl Node {
    fn new() -> Self {
        Self {
            children: HashMap::new(),
            is_end: false,
            number: 0,
        }
    }

    fn insert(&mut self, s: &str, number: u32) {
        let mut node: &mut Node = self;
        for c in s.chars() {
            node = node.children.entry(c).or_insert(Node::new());
        }
        node.is_end = true;
        node.number = number;
    }

    fn find_first_match(&self, s: &str) -> Option<u32> {
        let mut current_node: &Node = self;

        for c in s.chars() {
            if let Some(child) = current_node.children.get(&c) {
                current_node = child;
            } else {
                break;
            }
        }
        if current_node.is_end {
            Some(current_node.number)
        } else {
            None
        }
    }
}

fn run(input: &str) -> u32 {
    // Your code goes here
    let mut sum_: u32 = 0;

    let mut number_search_tree: Node = Node::new();

    let word_numbers: [&str; 9] = [
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];
    let numbers: [char; 9] = ['1', '2', '3', '4', '5', '6', '7', '8', '9'];

    for i in 0..9 {
        number_search_tree.insert(word_numbers[i], numbers[i].to_digit(10).unwrap());
        number_search_tree.insert(
            numbers[i].to_string().as_str(),
            numbers[i].to_digit(10).unwrap(),
        );
    }

    for line in input.lines() {
        let mut is_first_set: bool = false;
        let mut first: u32 = 0;
        let mut last: u32 = 0;
        for i in 0..line.len() {
            let n_ = number_search_tree.find_first_match(&line[i..]);
            if n_.is_some() {
                if !is_first_set {
                    first = n_.unwrap();
                    is_first_set = true;
                }
                last = n_.unwrap();
            }
        }
        sum_ += first * 10 + last;
    }
    sum_
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn run_test() {
        assert_eq!(run("Test example"), 0)
    }
}
